sage/gen_split_lambda_constants.sage

```diff
@@ -0,0 +1,97 @@
+load("secp256k1_params.sage")
+
+def inf_norm(v):
+    """Returns the infinity norm of a vector."""
+    return max(map(abs, v))
+
+def gauss_reduction(i1, i2):
+    v1, v2 = i1.copy(), i2.copy()
+    while True:
+        if inf_norm(v2) < inf_norm(v1):
+            v1, v2 = v2, v1
+        m = round( (v1[0]*v2[0] + v1[1]*v2[1]) / inf_norm(v1)**2 )
+        if m == 0:
+            return (v1, v2)
+        v2[0] = v2[0] - m*v1[0]
+        v2[1] = v2[1] - m*v1[1]
+
+def find_split_constants_gauss():
+    """Find constants for secp256k1_scalar_split_lamdba using gauss reduction."""
+    (v11, v12), (v21, v22) = gauss_reduction([0, N], [1, int(LAMBDA)])
+
+    # We use related vectors in secp256k1_scalar_split_lambda
+    A1, B1 = -v21, -v11
+    A2, B2 = v22, -v21
+
+    return (A1, B1, A2, B2)
+
+def find_split_constants_explicit_tof():
+    """Find constants for secp256k1_scalar_split_lamdba using the trace of Frobenius.
+
+    See Benjamin Smith: "Easy scalar decompositions for efficient scalar multiplication on
+    elliptic curves and genus 2 Jacobians" (https://eprint.iacr.org/2013/672), Example 2
+    """
+    assert P % 3 == 1 # The paper says P % 3 == 2 but that appears to be a mistake, see [10].
+    assert C.j_invariant() == 0
+
+    t = C.trace_of_frobenius()
+
+    c = Integer(sqrt((4*P - t**2)/3))
+    A1 = Integer((t - c)/2 - 1)
+    B1 = c
+
+    A2 = Integer((t + c)/2 - 1)
+    B2 = Integer(1 - (t - c)/2)
+
+    # We use a negated b values in secp256k1_scalar_split_lambda
+    B1, B2 = -B1, -B2
+
+    return A1, B1, A2, B2
+
+A1, B1, A2, B2 = find_split_constants_explicit_tof()
+
+# For extra fun, use an independent method to recompute the constants
+assert (A1, B1, A2, B2) == find_split_constants_gauss()
+
+assert A1*B2 - B1*A2 == N
```

Comment:
If you want to add a comment, I think it would be that this determinant check is verifying that the vectors, um, bound? surround? a fundamental domain (of the ring homomorphism from Z[-1/2+i*sqrt(3)/2] to Zn).

Reply:
If you move this assertion below the next two you could argue that this is proving that the two vectors are not only in the kernel but generate the kernel as well and phrase the comment that way.

(note the determinant also proves the vectors are linearly independent by being non-zero).

Reply:
> If you want to add a comment, I think it would be that this determinant check is verifying that the vectors, um, bound? surround? a fundamental domain (of the ring homomorphism from Z[-1/2+i*sqrt(3)/2] to Zn).

Can you suggest a wording? I hadn't heard "fundamental domain" before in my life.

> If you move this assertion below the next two you could argue that this is proving that the two vectors are not only in the kernel but generate the kernel as well and phrase the comment that way.
> 
> (note the determinant also proves the vectors are linearly independent by being non-zero).

Indeed, that's easier. But we could still say why it's exactly n.

Reply:
I don't know.  Maybe I'd say, "Check that he parallelogram generated by <A1, A2> and <B1,B2> is a fundamental domain that contains exactly N points."